File: back-end-application/order-service/src/main/java/be/technifutur/order/business/service/OrderService.java
package be.technifutur.order.business.service;

import be.technifutur.order.business.mapper.OrderMapper;
import be.technifutur.order.model.dto.OrderDTO;
import be.technifutur.order.model.entity.Order;
import be.technifutur.order.model.form.OrderForm;
import be.technifutur.order.repository.GameRepository;
import be.technifutur.order.repository.OrderRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@AllArgsConstructor
public class OrderService {

    private final OrderRepository orderRepository;
    private final GameRepository gameRepository;
    private final OrderMapper mapper;

    // CREATE NEW ORDER
    public OrderDTO placeOrder(UUID userRef, OrderForm form) {
        Order entity = mapper.formToEntity(form);
        orderRepository.findByUserRef(userRef).add(entity);
        return mapper.entityToDTO(entity);
    }

    // READ ORDERS LIST OF A SPECIFIC USER
    public List<OrderDTO> getOrdersByUser(UUID userRef) {
        return orderRepository.findByUserRef(userRef)
                .stream()
                .map(mapper::entityToDTO)
                .toList();
    }

    // READ ONE ORDER OF A SPECIFIC USER
    public OrderDTO getOneUserOrderByOrderId(UUID userRef, Long orderId) {
        Order orderToGet = orderRepository.findByUserRefAndId(userRef, orderId);
        return mapper.entityToDTO(orderToGet);
    }

    // UPDATE ONE ORDER OF A SPECIFIC USER
//    public OrderDTO updateUserOrderByOrderId(UUID userRef, UUID gameRef, Long orderId, int quantity) {
//        Order orderToUpdate = orderRepository.findByUserRefAndId(userRef, orderId);
//    }

    // DELETE ONE USER ORDER BY ID
    public OrderDTO deleteUserOrderById(UUID userRef, Long orderId) {
        Order orderToDelete = orderRepository.deleteByUserRefAndId(userRef, orderId);
        return mapper.entityToDTO(orderToDelete);
    }

}

File: front-end-application/tp4-mist-angular/src/app/_services/order.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Order, OrderForm } from '../_models/order.model';

@Injectable({
  providedIn: 'root'
})
export class OrderService {

  private BASE_URL: string = 'http://localhost:8080/orders';

  constructor(private http: HttpClient) { }

  public placeOrder(form: OrderForm): Observable<Order> {
    return this.http.post<Order>(`${this.BASE_URL}/add`, form);
  }

  public getOrdersByUser(): Observable<Array<Order>> {
    return this.http.get<Array<Order>>(`${this.BASE_URL}`);
  }

  public getOneUserOrderByOrderId(orderId: number): Observable<Order> {
    return this.http.get<Order>(`${this.BASE_URL}/${orderId}`);
  }

  public deleteUserOrderById(orderId: number): Observable<Order> {
    return this.http.delete<Order>(`${this.BASE_URL}/${orderId}`);
  }
  
}
